src/Xamarin.Android.Build.Tasks/Linker/MonoDroid.Tuner/FixAbstractMethodsStep.cs

```diff
@@ -19,43 +19,79 @@ namespace MonoDroid.Tuner
 	/// <summary>
 	/// NOTE: this step is subclassed so it can be called directly from Xamarin.Android.Build.Tasks
 	/// </summary>
-	public class FixAbstractMethodsStep : BaseStep
+	public class FixAbstractMethodsStep :
+#if NET5_LINKER
+	BaseMarkHandler
+#else   // !NET5_LINKER
+	BaseStep
+#endif  // !NET5_LINKER
 	{
-		readonly TypeDefinitionCache cache;
+
+#if NET5_LINKER
+		public override void Initialize (LinkContext context, MarkContext markContext)
+		{
+			base.Initialize (context, markContext);
+			markContext.RegisterMarkTypeAction (type => ProcessType (type));
+		}
+#else   // !NET5_LINKER
+		readonly TypeDefinitionCache resolver;
 
 		public FixAbstractMethodsStep (TypeDefinitionCache cache)
 		{
-			this.cache = cache;
+			resolver = cache;
 		}
+#endif  // !NET5_LINKER
 
-		protected override void ProcessAssembly (AssemblyDefinition assembly)
+		bool CheckShouldProcessAssembly (AssemblyDefinition assembly)
 		{
 			if (!Annotations.HasAction (assembly))
 				Annotations.SetAction (assembly, AssemblyAction.Skip);
 
 			if (IsProductOrSdkAssembly (assembly))
-				return;
+				return false;
 
 #if !NET5_LINKER
 			CheckAppDomainUsageUnconditional (assembly, (string msg) => Context.LogMessage (MessageImportance.High, msg));
-#endif
+#endif  // !NET5_LINKER
+
+			return assembly.MainModule.HasTypeReference ("Java.Lang.Object");
```

Comment:
*This* concerns me, and I don't remember seeing it in a previous iteration of this PR. See also:

  * https://github.com/xamarin/xamarin-android/pull/5965
  * https://github.com/xamarin/xamarin-android/pull/4225#issuecomment-583139797

There is *no* requirement that an assembly which contains types which *need* to have abstract methods "fixed up" *also* contain a type reference to `Java.Lang.Object`, much less a reference to `Mono.Android.dll`!  Such is the glory of transitive references.

I'm fairly sure this check will *break* existing code.

Reply:
Your concern makes sense. This is an existing check (and was also present in previous iterations) - I moved it because this step now runs for marked types, instead of for all assemblies. I just filed https://github.com/xamarin/xamarin-android/issues/5968 about it - could this be addressed separately? I'm trying to minimize functional changes in this PR.

Reply:
Sorry, I didn't realize this check was originally there; I didn't see the movement from `FixAbstractMethodsUnconditional()`. :frowning_face: